### PR TITLE
Fix mxr writes and increase read limits

### PR DESCRIPTION
**Pull request by TedThemistokleous**
## Motivation

Fix issue seen when dealing with large models with parameters that have dynamic shapes. This fixes an underflow issue seen when working on a customer model set. had a few things flagged by cursor with regards to reading limits.

 ~also seeing that we're encoding floats incorrectly (32 instead of 64 bits) as part of the mxr process wasting a bit of space.~

## Technical Details

- Handle underflow generated when we get binary data that's in an array of zero - this was the cause of `insufficient bytes` errors
~- Handle float32 as float -> saves half the size for the final mxr for floats that were being packed as float64 erroneously~
- Increase limit for reading of large models/mxrs -> should allows to avoid hitting msgpack read limits (future proofing)

## Changelog Category

- - [ ] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [x] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

### CODE REVIEW

File: src/msgpack.cpp [hunk 1]
```diff
@@ -244,7 +248,17 @@ std::vector<char> to_msgpack(const value& v)
...
-    msgpack::object_handle oh = msgpack::unpack(buffer, size);
+    // Use higher limits for deeply nested structures (e.g., shapes with tuple_type containing sub_shapes)
+    msgpack::unpack_limit limits(
+        0xffffffff,  // array - no limit
+        0xffffffff,  // map - no limit
+        0xffffffff,  // str - no limit
+        0xffffffff,  // bin - no limit
+        0xffffffff,  // ext - no limit
+        0xffffffff   // depth - allow very deep nesting for complex models
+    );
```
Comment (pfultz2) [inline src/msgpack.cpp]: The default limits are already set to `0xffffffff`, so I dont think this is needed: https://github.com/msgpack/msgpack-c/wiki/v2_0_cpp_unpacker#limit-size-of-elements

File: src/msgpack.cpp [hunk 2]
```diff
@@ -81,6 +83,8 @@ MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS)
                 break;
             }
             case msgpack::type::FLOAT32:
+                v = o.as<float>();
+                break;
```
Comment (pfultz2) [inline src/msgpack.cpp]: This is not needed. The value class always stores it as double.

Comment (TedThemistokleous) [reply src/msgpack.cpp]: Oh okay. Ill add a comment to clarify it and revert this change. It looked odd, and when I tried this, MXR as smaller. Is there a reason why value stores it as double and not  float32 then? Is that something we want to fix or need to be fixed?

File: src/msgpack.cpp [hunk 3]
```diff
@@ -34,6 +34,8 @@ constexpr std::size_t msgpack_size_limit = std::numeric_limits<uint32_t>::max()
 template <class Range>
 static std::size_t msgpack_chunk_size(const Range& r)
 {
+    if(r.size() == 0)
+        return 1;  // Even empty data needs one chunk
```
Comment (pfultz2) [inline src/msgpack.cpp]: Can you add a unit test for this?

Also, I think it would be better to put a lower bound for this in the return below: `return 1 + (std::max<std::size_t>(1, r.size()) - 1) / msgpack_size_limit`

File: src/msgpack.cpp (hunk 2 above)
Comment (TedThemistokleous) [reply src/msgpack.cpp]: Removed this

File: src/msgpack.cpp (hunk 3 above)
Comment (TedThemistokleous) [reply src/msgpack.cpp]: That wont work since size_t - 1 rolls over still (underflow) since its unsigned, and we'll still go with the larger value, regardless of the 1. I don't see a case where size() somehow decays to an signed value nor does that make sense. Aren't we bound here by size_t?

File: src/msgpack.cpp (hunk 1 above)
Comment (TedThemistokleous) [reply src/msgpack.cpp]: Sure I'll remove it then.

Comment (TedThemistokleous) [reply src/msgpack.cpp]: Removed

**Comment by TedThemistokleous** [comment]
@pfultz2 added a simple testcase where we create an empty array. Should cover us with this case

File: src/msgpack.cpp (hunk 3 above)
Comment (TedThemistokleous) [reply src/msgpack.cpp]: Tried this with the unit test I created. This doesn't seem to work. 

```
/code/AMDMIGraphX/test/msgpack.cpp:227:
    FAILED: migraphx::to_msgpack(bin) == msgpack_buffer(bin) [ { �, �, } == { �, �, } ]
[  FAILED  ] test_msgpack_binary_empty (9.98415ms): Test failure
```

Comment (kahmed10) [reply src/msgpack.cpp]: try `return std::max<std::size_t>(1, (r.size() + msgpack_size_limit - 1) / msgpack_size_limit);`

Comment (pfultz2) [reply src/msgpack.cpp]: > That wont work since size_t - 1 rolls over still (underflow) since its unsigned

It doesnt do `r.size() - 1`, it does `max(1, r.size()) - 1`, so there is no way for it to underflow since the smallest value is 1.

Comment (pfultz2) [reply src/msgpack.cpp]: > Tried this with the unit test I created. This doesn't seem to work.

You need to do the same guard in `write_msgpack` function in the tests as well.

File: test/msgpack.cpp [hunk 4]
```diff
@@ -220,5 +220,12 @@ TEST_CASE(test_msgpack_binary2)
     });
     EXPECT(migraphx::to_msgpack(bin) == msgpack_buffer(bin));
 }
+
+TEST_CASE(test_msgpack_binary_empty)
+{
+    migraphx::value::binary bin{};
+    EXPECT(migraphx::to_msgpack(bin) == msgpack_buffer(bin));
```
Comment (pfultz2) [inline test/msgpack.cpp:227]: You could just check the roundtrip would be the same.

File: test/msgpack.cpp [hunk 5]
```diff
@@ -220,5 +220,12 @@ TEST_CASE(test_msgpack_binary2)
     });
     EXPECT(migraphx::to_msgpack(bin) == msgpack_buffer(bin));
 }
+
+TEST_CASE(test_msgpack_binary_empty)
+{
+    migraphx::value::binary bin{};
```
Comment (pfultz2) [inline test/msgpack.cpp:226]: This should be declared as `migraphx::value bin = migraphx::value::binary{}`, so it gets declared as a `value` type which makes it easier to roundtrip check it.

Comment (TedThemistokleous) [reply test/msgpack.cpp:226]: Seems like its worth another test for that then no? simple empty check. If roundtrip fails but this doesn't that would signal something else is broken.

File: src/msgpack.cpp (hunk 3 above)
Comment (TedThemistokleous) [reply src/msgpack.cpp]: In the test? I dont see what you mean with that. We're determining chunk size and using that in the msgpack.cpp but not in the tests

File: test/msgpack.cpp (hunk 5 above)
Comment (TedThemistokleous) [reply test/msgpack.cpp:226]: I add another test below for the specific case you mentioned. Should catch roundtrip errors

File: src/msgpack.cpp (hunk 3 above)
Comment (pfultz2) [reply src/msgpack.cpp]: We do but we only test 2 chunks, i thought the test fails with `write_msgpack` because this needs to be updated to handle empty size:

https://github.com/ROCm/AMDMIGraphX/blob/develop/test/msgpack.cpp#L40

Comment (TedThemistokleous) [reply src/msgpack.cpp]: Nope, I'm an idiot. Didn't see the extra bracket and completely misinterpreted what you were saying. First line works. 

No change needed on test. Updated another test to do roundtrip to but taking in the empty binary data and handle that correctly.

File: test/msgpack.cpp (hunk 5 above)
Comment (TedThemistokleous) [reply test/msgpack.cpp:226]: Yeah cleaned that test up. We should test both the to/from msgpack from the binary value, similar to the empty value test we have but assigning binary data to ensure that doesn't give us surprises. Called it `test_msgpack_binary_roundtrip_empty`

File: test/msgpack.cpp (hunk 4 above)
Comment (TedThemistokleous) [reply test/msgpack.cpp:227]: Done

**Review by kahmed10 — APPROVED** [review]
